utils/wasm-builder/src/lib.rs

```diff
@@ -172,73 +176,118 @@ fn copy_file_if_changed(src: PathBuf, dst: PathBuf) {
+		// If no command before provided us with a cargo that supports our Substrate wasm env, we
+		// try to search one with rustup. If that fails as well, we return the default cargo and let
+		// the prequisities check fail.
+		get_rustup_command().unwrap_or(default_cargo)
 	}
 }
 
-/// Get a nightly from rustup. If `selected` is `Some(_)`, a `CargoCommand` using the given
-/// nightly is returned.
-fn get_rustup_nightly(selected: Option<String>) -> Option<CargoCommand> {
+/// Get the newest rustup command that supports our Substrate wasm env.
+///
+/// Stable versions are always favored over nightly versions even if the nightly versions are
+/// newer.
+fn get_rustup_command() -> Option<CargoCommand> {
 	let host = format!("-{}", env::var("HOST").expect("`HOST` is always set by cargo"));
 
-	let version = match selected {
-		Some(selected) => selected,
-		None => {
-			let output = Command::new("rustup").args(&["toolchain", "list"]).output().ok()?.stdout;
-			let lines = output.as_slice().lines();
-
-			let mut latest_nightly = None;
-			for line in lines.filter_map(|l| l.ok()) {
-				if line.starts_with("nightly-") && line.ends_with(&host) {
-					// Rustup prints them sorted
-					latest_nightly = Some(line.clone());
-				}
+	let output = Command::new("rustup").args(&["toolchain", "list"]).output().ok()?.stdout;
+	let lines = output.as_slice().lines();
+
+	let mut found_version: Option<(String, Version)> = None;
+	for line in lines.filter_map(|l| l.ok()) {
+		let rustup_version = line.trim_end_matches(&host);
+
+		let cmd = CargoCommand::new_with_args("rustup", &["run", &rustup_version, "cargo"]);
+
+		if !cmd.supports_substrate_wasm_env() {
+			continue
+		}
+
+		let Some(cargo_version) = cmd.version() else { continue; };
+
+		if let Some((rustup_version_other, cargo_version_other)) = &mut found_version {
+			let is_older = cargo_version_other.is_older(&cargo_version);
+
+			// Nightly should not overwrite a stable version.
+			if cargo_version_other.is_stable() && cargo_version.is_nightly {
+				continue
 			}
 
-			latest_nightly?.trim_end_matches(&host).into()
-		},
-	};
+			// Stable versions overwrite nightly versions or otherwise just check which one is
+			// older.
+			if cargo_version_other.is_nightly && cargo_version.is_stable() || is_older {
+				*rustup_version_other = rustup_version.into();
+				*cargo_version_other = cargo_version;
+			}
+		} else {
+			found_version = Some((rustup_version.into(), cargo_version));
+		}
+	}
 
+	let version = found_version?.0;
 	Some(CargoCommand::new_with_args("rustup", &["run", &version, "cargo"]))
 }
 
+/// The version of rustc/cargo.
+#[derive(Clone, Copy, Debug)]
+struct Version {
+	pub major: u32,
+	pub minor: u32,
+	pub patch: u32,
+	pub is_nightly: bool,
+}
+
+impl Version {
+	/// Returns if `self` is older than `other`.
+	fn is_older(&self, other: &Self) -> bool {
+		self.major < other.major || self.minor < other.minor || self.patch < other.patch
```

Comment:
3.2.1 - 2.4.1 will report `true` even though 2.4.1 is older. I think this should be smth like

```
self.major < other.major || (self.major == other.major && self.minor < other.minor) || (self.major == other.major && self.minor == other.minor && self.patch < other.patch)
```

Reply:
This should also compare dates if the versions are equal, as the versions (for nightlies) can quite easily be the same e.g.:

```
$ rustup run nightly-2023-02-14-x86_64-unknown-linux-gnu rustc --version
rustc 1.69.0-nightly (065852def 2023-02-13)
$ rustup run nightly-2023-01-24-x86_64-unknown-linux-gnu rustc --version
rustc 1.69.0-nightly (c8e6a9e8b 2023-01-23)
```